Completely replace perfect-scrollbar with SimpleBar

andersk opened the pull request:
perfect-scrollbar replaces both the appearance and the behavior of the scrollbar, and its emulated behavior will never feel native on most platforms.  SimpleBar customizes the appearance while preserving the native behavior.

**Testing Plan:** I found and fixed a number of obvious issues just by clicking around, but there could easily be more. We should make sure to carefully test every element that uses `set_up_scrollbar`.

Fixes: #12038

timabbott commented on static/js/ui_init.js:
```diff
@@ -90,7 +90,7 @@ exports.initialize_kitchen_sink_stuff = function () {
     // element is already at the top or bottom.  Otherwise we get a
     // new scroll event on the parent (?).
-    $('.modal-body, .scrolling_list, input, textarea').on('wheel', function (e) {
+    $('input, textarea').on('wheel', function (e) {
```
What's the story with this change?

andersk replied:
With this code blocking wheel events, the scrollbars weren’t working at all in some cases. I optimistically assumed that the issue this code was trying to work around was likely specific to perfect-scrollbar, so I figured it’s worth trying to remove it under the Less Code Is Better principle. If the issue reappears, we’ll need to think about it more.

timabbott replied:
I spelunked this a bit, and it's ca2b48e1ee754636a00490460493f1adb6d2dc04 and 45e515cc20fbeee73b91497fdd5005f2c06cde3b that introduce this.  

I think that predates perfect-scrollbar (1eae686443ed3f8c58156af6885a8ff9a4a683ac) by many months.  I would test scrolling in "keyboard shortcuts", the sidebars, and various settings views, and make sure they don't move the scroll position in the message feed; if that works, I'm happy to assume that the ancient bug that motivated this no longer exist.

(As a future note, it's best to give this sort of reasoning in a commit message or comments you add on the PR after opening it rather than just not mentioning it, since a reviewer is definitely going to ask why this random thing was removed).

andersk replied:
Upon further investigation, I think neither the current behavior nor my modified behavior were fully correct, so I’ve made the changes necessary to reimplement the current behavior for now.

timabbott commented on static/js/ui.js:
```diff
@@ -13,34 +15,37 @@ exports.replace_emoji_with_text = function (element) {
+        element.SimpleBar.getScrollElement().scrollTop = 0;
+        element.SimpleBar.recalculate();
+    } else {
+        element.scrollTop = 0;
```
When does this else clause come up?  Or is this just trying to port the existing logic?

andersk replied:
Porting existing logic, yes, but it comes up opening “Manage streams” for some reason.

timabbott commented on static/js/ui.js:
```diff
@@ -13,34 +15,37 @@ exports.replace_emoji_with_text = function (element) {
-    element[0].perfectScrollbar = perfectScrollbar;
+exports.set_up_scrollbar = function (element_selector) {
+    new SimpleBar(element_selector.expectOne()[0]);
+};
```
We should consider preserving the settings we had before.  From https://github.com/Grsmto/simplebar#options, should we:

* Set `scrollbarMinSize` should be 40 to match `minScrollbarLength` above?
* Include an option to `set_up_scrollbar` to set `forceVisible`?  (I think we have at least one place we'll want to use this).
* Do something to emulate the `suppressScrollX` feature (https://github.com/Grsmto/simplebar#apply-scroll-vertically-only)?  I guess mostly we'll need to find the places that need this in QA.

timabbott replied:
(The place we wanted forceVisible was on the right panel of "settings")

andersk replied:
* The SimpleBar default `scrollbarMinSize = 25` seems fine to me. (perfect-scrollbar had no default `minScrollbarLength`.)
* `forceVisible` is easy to add later, so let’s do one thing at a time.
* We don’t need to do anything for `suppressScrollX` if the CSS is correct, so yeah, let’s make the CSS correct if it isn’t already.

timabbott replied:
ok

timabbott commented:
I posted a few questions, and am about to do some QA.  If that goes well, we can test-deploy on chat.zulip.org today.  While we're thinking about this, we should also review this list: https://github.com/zulip/zulip/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+scrollbar to see if we can clean out other scrollbar-related issues while we're at it.

timabbott commented:
Found this while doing QA (click a stream inside "manage streams"):

![image](https://user-images.githubusercontent.com/2746074/53912403-b4961d00-400d-11e9-9174-db8707636c44.png)

timabbott commented:
I think this will as a side effect fix another bug I noticed today, so probably worth giving this some priority.

timabbott commented:
Actually, make that 2 bugs that I discovered yesterday.

andersk commented:
This should be ready for a second round of QA.

rishig commented:
Did a quick partial round of QA. I didn't do a diff with the current behavior, so possible some of these are existing issues. 

**Settings**
* Scrollbar is too narrow / hard to click on. I would make it wider (and correspondingly adjust the color to be a little lighter).
* I think the rightmost scrollbar should just be persistent (on tabs where we need it), and not fade out on a timer. 
* I think the left scrollbar could be persistent. But another option: when you're hovering over the left sidebar, we show it, and when you're not, we remove it (again, no fade out on timer). 
* The left scrollbar goes up too high (it goes to the top of the modal, rather than the top of "Your account").

* Ideally the scrollbar darken when you click on it (low priority if hard)
* On "Your bots", if you create a bunch of bots, we correctly show the scroll bar on the "Active bots" subtab, and incorrectly show the scrollbar on "Add a new bot". (low priority if hard)
* On "Users", if you make the window short we incorrectly show a scrollbar when there isn't anything to scroll to (low priority if hard)

(did not check the attachments tab)

**Right sidebar**
ah, a few similar comments as with Settings. The scrollbar is too narrow / hard to click on, and also there should be a persistent scrollbar that doesn't fade out. I like how it's currently done on master; not sure if that's possible to achieve with simplebar. If not, we can think of a workaround.

**Emoji picker**
Scrollbar seems to be missing for me.

timabbott commented:
> I think the rightmost scrollbar should just be persistent (on tabs where we need it), and not fade out on a timer.

We discussed this and it is out of scope for this PR -- there's a flag for this in simplebar, we're just not dealing with feature changes.

andersk commented:
> **Settings**
> 
> * Scrollbar is too narrow / hard to click on. I would make it wider (and correspondingly adjust the color to be a little lighter).

I think we just need to override `.simplebar-scrollbar:hover { width }` in CSS for this.

> * I think the rightmost scrollbar should just be persistent (on tabs where we need it), and not fade out on a timer.

Addressed by Tim above. Actually the timeout is a change from perfect-scrollbar. We can disable `autoHide`, or we can lengthen `timeout`, but I don’t think we want either of those on *every* SimpleBar, so I’m inclined to leave this for future work if it isn’t critical.

> * I think the left scrollbar could be persistent. But another option: when you're hovering over the left sidebar, we show it, and when you're not, we remove it (again, no fade out on timer).

See above.

> * The left scrollbar goes up too high (it goes to the top of the modal, rather than the top of "Your account").

Matches old behavior.

> * Ideally the scrollbar darken when you click on it (low priority if hard)

Old behavior has no darkening on click, and weird inconsistent flashing between dark and light while dragging.

> * On "Your bots", if you create a bunch of bots, we correctly show the scroll bar on the "Active bots" subtab, and incorrectly show the scrollbar on "Add a new bot". (low priority if hard)

Matches old behavior (missing `update_scrollbar` calls probably?).

> * On "Users", if you make the window short we incorrectly show a scrollbar when there isn't anything to scroll to (low priority if hard)

I can’t seem to reproduce?

> (did not check the attachments tab)
> 
> **Right sidebar**
> ah, a few similar comments as with Settings. The scrollbar is too narrow / hard to click on, and also there should be a persistent scrollbar that doesn't fade out. I like how it's currently done on master; not sure if that's possible to achieve with simplebar. If not, we can think of a workaround.

See above.

> **Emoji picker**
> Scrollbar seems to be missing for me.

Yep, that’s a bug. Added an easy CSS fix. (SimpleBar requires `overflow: auto` or `overflow-y: auto`.)

timabbott commented:
OK, I test-deployed this on chat.zulip.org for feedback.  I think we should start building on top of this branch fixes for any scrollbar things we want to change, under the assumption that we're merging this migration:

* There's a bug on chat.zulip.org where the right-sidebar scrollbar is the entire height of the sidebar.  I believe this is likely a bad interaction with the right sidebar paginated rendering strategy we use.  @showell may be able to help debug.
* We should go ahead and add the persistent scrollbars in the places we've mentioned in this thread in a new commit, since it's likely easy and will probably help avoid distractions in the discussion
* Worth trying to fix the "update_scrollbar" issue with bots.
* Worth doing a pass on editing the scrollbar's styling

And then we can iteratively test-deploy this on chat.zulip.org over the next couple days.

andersk commented:
I did some restyling to match the widen-on-hover behavior of perfect-scrollbar and lighten the color to a gray that works in day and night modes.

Are there any other reproducible regressions?

andersk commented:
Just to keep this updated with current state, I found and fixed what I think was causing the right sidebar problems (missing `ui.get_scroll_element` calls).

timabbott commented:
OK, doing a new deploy on chat.zulip.org now.

timabbott commented:
So our last deployment of this was on chat.zulip.org for like a week (just pulled it off last night because I didn't want to deal with a merge conflict; re-deploying to chat.zulip.org now).  Do we have other open issues with this PR that should block merging?  I think in any case it'd be useful to add additional commits on top to add permanent scrollbars in the places we want them as discussed in @rishig's comments above.  

@rishig FYI.

rishig commented:
There's https://chat.zulip.org/#narrow/stream/9-issues/topic/scroll.20bar.20on.20USERS from recent memory.

timabbott commented:
Hrm, OK, hopefully we'll get a quick merge for that upstream.

andersk commented:
The next simplebar release should make all the `recalculate`/`update_scrollbar` calls unnecessary (Grsmto/simplebar#282), so that’s also worth waiting for.

Grsmto commented:
Hey, just saw you guys are implementing SimpleBar. FYI I'll be doing a release in the next few days. As it's a breaking change (due to DOM structure change, shouldn't impact most users) I'm stacking a few bug fixes with it so it takes a big longer!

TheCelestial25 commented:
I was working on this scrollbar as I worked on #11694. I wanted to know if it is required to set autoHide to false regards to the issue I mentioned or the current behavior is fine.
I also tried to fix the buggy behavior in user_profile_modal, towards which I would like to draw your attention. There is an issue with the rendering of this scrollbar, the scrollbar is not present and the margin/padding on the right side (where the scrollbar should be) vanishes.
![Capture](https://user-images.githubusercontent.com/37958393/56453460-86bc2c00-6360-11e9-959b-e843ff266bb5.JPG)
The margin/padding is fine for the browser rendered scrollbar through.
![Capture](https://user-images.githubusercontent.com/37958393/56453653-3befe380-6363-11e9-9233-3a7d43e16517.JPG)

andersk commented:
@TheCelestial25 Thanks for the report. I’ve tracked this down to a SimpleBar issue that I filed as Grsmto/simplebar#305.

timabbott commented:
I'm testing the first 3 commits to make sure /help/ works, and then should be able to merge those.

timabbott commented:
Merged just the first 2 commits, because the simplebar upgrade seems to not have correct scroll-reset behavior when switching articles on /help/.

andersk commented:
> Merged just the first 2 commits, because the simplebar upgrade seems to not have correct scroll-reset behavior when switching articles on /help/.

This is a regression in SimpleBar.getScrollElement() between 4.0.0-alpha.3 and .4, filed as Grsmto/simplebar#308.

timabbott commented:
I merged the simplebar upgrade commit, yay!  I think @rishig had a couple things he noticed doing a QA pass of the other commit on chat.zulip.org that I'll mention here:

> Two things I noticed:
> * it's a little disconcerting that clicking on the scrollbar doesn't darken it, partly because the non-simplebar ones do darken. (Also on /help, it's on a non-white background so it's not that easy to see if not darkened.)
> * on the settings pages, if you click on the scrollbar, drag it down, and then release your click when the mouse is outside the settings modal (e.g. below it or to the right), it closes the settings modal. I don't know if this is an existing thing or a regression, but I ran into it a bunch of times when testing even after knowing the behavior.

I can reproduce the second thing as not a regression, but it is potentially worth looking into as a follow up.

So possibly we get to merge the migration commit too soon.

(Does simplebar support that sort of darken-on-click/hover as an option?)

andersk commented:
> * it's a little disconcerting that clicking on the scrollbar doesn't darken it, partly because the non-simplebar ones do darken. (Also on /help, it's on a non-white background so it's not that easy to see if not darkened.)

(“Non-simplebar” = perfect-scrollbar?)

I think we could do this with some custom CSS, but we’d need to be careful because naïvely darkening them would make them harder rather than easier to see in night mode.

> * on the settings pages, if you click on the scrollbar, drag it down, and then release your click when the mouse is outside the settings modal (e.g. below it or to the right), it closes the settings modal. I don't know if this is an existing thing or a regression, but I ran into it a bunch of times when testing even after knowing the behavior.

PR sent to Grsmto/simplebar#312.

rishig commented:
> PR sent to Grsmto/simplebar#312.

Thanks!

> (“Non-simplebar” = perfect-scrollbar?)

I guess I just meant the browser scrollbar, at least on Chrome.

andersk commented:
> * on the settings pages, if you click on the scrollbar, drag it down, and then release your click when the mouse is outside the settings modal (e.g. below it or to the right), it closes the settings modal.

Further upstream PR: Grsmto/simplebar#317. (Still not a regression though.)

timabbott commented:
Merged, huge thanks for your work on this @andersk!  I think we can fix-forward any further issues from here.

We should open follow-up issues for the remaining items discussed above that don't have their own issue (or maybe just a master scrollbar follow-ups issue with little checkboxes?).  @andersk can you take care of that?  I don't have a great handle on which of the things @rishig and I mentioned above are fixed in the current state.

Also there's a bunch of random scrollbar issues we should check if are fixed and if not, consider clearing out while we're thinking about scrollbars:
* https://github.com/zulip/zulip/issues/11694
* https://github.com/zulip/zulip/issues/10544
* https://github.com/zulip/zulip/issues/10421
* https://github.com/zulip/zulip/issues/8851
* https://github.com/zulip/zulip/issues/9039
* Possibly most other things in https://github.com/zulip/zulip/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+scrollbar, though some of them are UI changes I'm not sure we want to do.
